### ISSUE TITLE
Support for Container/Kubernetes environments?

### ISSUE THREAD

**Issue by christianhuening** [tonarino/innernet#42]
Hi, we just found innernet and so far like it a lot. We're wondering whether you already have any plans or setups to run this in Kubernetes environments? We are thinking about using innernet to hide our development and production clusters within a private company network. 
We would go ahead and produce a dockerfile and a helm chart to try it, but just wanted to ask first, whether it has been done already somewhere.

**Comment by mcginty** [comment]
Hey @christianhuening! Thanks I'm glad you like it so far! We don't use kubernetes internally so I haven't tested it, *but* it should work just fine in a Docker container if the host has the WireGuard kernel extension available, and if not you can instead run `wireguard-go` inside the container.

Something like https://github.com/cmulk/wireguard-docker (or https://github.com/masipcat/wireguard-go-docker for the second option) might be a good starting point for making your `Dockerfile` - in the next week or so I plan on adding better automated testing to this repo using either Docker or QEMU too, so there may be more news on that front in a bit.

Long story short, it should work as long as you're comfortable with adding the additional necessary `NET_ADMIN` capability for creating network interfaces on the host.

**Comment by alesti** [comment]
Hi, we (Christian is one of my coworkers) appreciate your fast responses - it makes us really happy and also confident that we will use innernet in future.

I will try to explain a little bit more our docker (or better kubernetes) specific problems, we try to handle:

We are on the way to hide our public, but non workload endpoints of our clusters (api server, alertmanager, grafana, metrics, ...)

We know tailscale, but is very expensive and we cannot run our own yellow pages server - we work in a high regulated sector and try to host most stuff by ourselfs.
We checked [pritunl](https://docs.pritunl.com/docs), which uses a mongodb as configuration data store and has openvpn and wireguard as vpn protocols. It has a nifty webinterface to configure anything. It is

**Comment by alesti** [comment]
Hi, we (@christianhuening  is one of my coworkers) appreciate your fast responses - it makes us really happy and also confident that we can use innernet in future.

I will try to explain a little bit more our docker (or better kubernetes) specific problems, we try to handle:

We are on the way to hide our non workload endpoints of our clusters (api server, monitoring, metrics, ...) from public.

We know tailscale, but it is very expensive and we cannot run our own control server - we work in a heavly regulated sector and try to host most of the at least security related stuff by ourselfs.

We checked [pritunl](https://docs.pritunl.com/docs), which uses a mongodb as configuration data store and has openvpn and wireguard as vpn protocols. It has a nifty webinterface to configure anything and also client software with gui and a nice way to send configs to new clients. **But** it still uses the outdated python 2.7, and it is currently not easy to run it dockerized in k8s as it needs to expose udp and tcp (wireguard uses the api/webinterface for key exchange thingies) in parallel, which is simply not possible in k8s today. We tried to run it as standalone docker on one of our nodes (which is possible), but it is really complicated to route traffic into the k8s pod network.

We found innernet, and it has a much better design for our needs - a small configuration database, easy to backup/restore, mesh appproach, we need only to expose one udp port.

As Christian wrote already, we will contribute our solution and helmchart, if we are confident that it works as expected.

I am on to build a innernet docker container build chain, and i will face the following docker/headless run problems. Some of them are addressed as tickets already by other users, some of them are already solved, which is really great! 

This list is just to get an idea, where we currently tinkering and thinking - no requests so far.

### Howto handle initial connect and key exchange in headless envs:

First idea: A init container slurps the inital config.toml from [vault](https://www.vaultproject.io/) (we have all secrets in it) and does `inn install config.toml` to get the working config and private key, writes the `vpn.conf` into a configmap resource.

The init container runs only if there is no `vpn.conf` as configmap available.

Problem with this: configmap is nothing secure, and a pod cannot write into a secret nor write into our vault.

- So we might accept that the private key is in danger, if there is the possibilty for a regular automatic key rotation (should be possible a this is done with the inital connect config).
- We do the initial connect manually and push the self generated key into vault (bad, as we plan to spin up k8s clusters with [gardener](https://github.com/gardener/gardener)).
- Innernet gets the command line options for install to slurp an external generated key pair (`wg genkey | tee privatekey | wg pubkey > publickey`) from the k8s secret (which is feed by vault)

(i think, i would prefer the last way)

### Key rotation 

It would be cool to use the already builtin key generation feature as command line option to rotate keys (also with provided key pair, see above)

### Default exposed udp port 

As writen in https://github.com/tonarino/innernet/issues/46, we need to know/define the port in front of the container deployment as we need to expose it right.

It would be nice to slurp the local_port from command line.

### Routing / side2side vpn

Even in the roadwarrior usecase (a dev needs to contact an api-server (cli tool kubectl) we need to provide routes to the peers as the pod running innernet is not the final destination.

We have a pure wireguard setup running which maintains one part of this in the per peer  AllowedIPs =  section and also in `PostUp = sleep 10 ; route add -net X.X.X.X/12 gw <peer address> dev wgX` of the wgX.conf on the server -  together with PostUp similar statements on peers side (routing, masquerading).
But in this setup the server is always central part of the network. 

Need to think (and possibly draw) about this...

This is addressed in https://github.com/tonarino/innernet/issues/22

### sudo is a problem..

... not for k8s, but for human peers. Looks like it is at least neccessary to write into /etc/hosts and up interfaces.
This is already addressed in another ticket (https://github.com/tonarino/innernet/issues/41)

### Find a `systemctl enable` equivalent for macos

Most of our users using macos and are not allowed to use sudo, so we need to explore macos' launchd to start innernet as service, this might also solve a part of the sudo problem. (i am not confident if launchd is still the way, macos does this, i will find out...)

### Connect innernet api to oauth2

We have some users, and we need to build a connection to our userdatabase. 


A lot stuff, dont get it wrong - thats not your list, but ours :-) 
I hope, some of you readers get a glimpse which points we want to get tackled.

**Comment by mcginty** [comment]
@christianhuening @alesti Thanks so much for that comment, and sorry for the delay in responding especially after your thanking us for prompt replies 😝. Even though I didn't reply, I've been focusing on improving the non-interactive functionality in innernet, and just opened a pull request (https://github.com/tonarino/innernet/pull/55) that has some a sample working `Dockerfile` for the client and server, using the new CLI flags I just added to the server and client binaries.

For simplicity I just use `docker cp` in my script, but uploading these configs and digesting them from something like `vault` should be a change that could work without further code changes.

> It would be cool to use the already builtin key generation feature as command line option to rotate keys (also with provided key pair, see above)

I'm very interested in implementing this feature - right now, the answer is simply to create a new peer invitation and migrate to that peer, but I find that clumsy and wasteful in the long-run. I'd be very interested in any proposals for how to support key rotation while maintaining a reasonable security model that can avoid peer keys being swapped out by unauthorized parties.

> We have a pure wireguard setup running which maintains one part of this in the per peer AllowedIPs = section (needed to remember myself that this are the networks a peer accepts traffic for, is this equivalent to innernets address = section?) and also in PostUp = sleep 10 ; route add -net X.X.X.X/12 gw <peer address> dev wgX of the wgX.conf on the server - together with PostUp similar statements on peers side (routing, masquerading).

Yep, innernet can be considered a special configuration of WireGuard where the `AllowedIPs` is always one address, which is why it's called `address` here. And yes, adding a site-to-site feature in innernet would change this fact - there would be some "bridge" peers that would have `AllowedIPs` that are more than one IP address.

> Find a `systemctl enable` equivalent for macos

Yeah, we'll be adding a launchd script shortly that could be manually managed or with a helper like `brew services`.

> Connect innernet api to oauth2

At present, I'm most interested in this being a separate project outside of the core `innernet` repository that can be a kind of "authentication bridge." I'm not yet convinced that it belongs in the main `innernet-server` binary, for example, but I think it will be an important project nonetheless.

**Comment by mcginty** [comment]
Closing this as the container tests have been working fine in testing!

**Comment by viq** [comment]
@alesti just jumping in to point out https://github.com/juanfont/headscale as something potentially interesting.
